multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,60 +95,110 @@ func NewMemberClusterSetReconciler(client client.Client,
 func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	clusterSet := &mcv1alpha2.ClusterSet{}
 	err := r.Get(ctx, req.NamespacedName, clusterSet)
-	r.commonAreaLock.Lock()
-	defer r.commonAreaLock.Unlock()
+	var clusterSetNotFound bool
+	var clusterSetCreated bool
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return ctrl.Result{}, err
 		}
-		klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
-		if r.remoteCommonArea != nil {
-			if err := r.deleteMemberClusterAnnounce(ctx); err != nil {
-				// MemberClusterAnnounce could be kept in the leader cluster, if antrea-mc-controller crashes after the failure.
-				// Leader cluster will delete the stale MemberClusterAnnounce with a garbage collection mechanism in this case.
-				return ctrl.Result{}, fmt.Errorf("failed to delete MemberClusterAnnounce in the leader cluster: %v", err)
+		clusterSetNotFound = true
+	}
+
+	processClusterSet := func() error {
+		r.commonAreaLock.Lock()
+		defer r.commonAreaLock.Unlock()
+
+		if clusterSetNotFound {
+			klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
+			if err := r.cleanUpResources(ctx); err != nil {
+				return err
 			}
-			r.remoteCommonArea.Stop()
-			r.remoteCommonArea = nil
-			r.clusterSetConfig = nil
-			r.clusterID = common.InvalidClusterID
-			r.clusterSetID = common.InvalidClusterSetID
+			return nil
 		}
-		return ctrl.Result{}, nil
-	}
 
-	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
+		klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
 
-	// Handle create or update
-	if r.clusterSetConfig == nil {
-		r.clusterID, err = common.GetClusterID(r.ClusterCalimCRDAvailable, req, r.Client, clusterSet)
-		if err != nil {
-			return ctrl.Result{}, err
+		// Handle create or update
+
+		newLeader := clusterSet.Spec.Leaders[0]
+
+		clusterSetCreated = r.clusterID != common.ClusterID(clusterSet.Spec.ClusterID) || r.clusterSetID != common.ClusterSetID(clusterSet.Name)
+		leaderChanged := r.installedLeader.clusterID != newLeader.ClusterID || r.installedLeader.serverUrl != newLeader.Server ||
+			r.installedLeader.secretName != newLeader.Secret
+		r.clusterSetConfig = clusterSet.DeepCopy()
```

Comment:
Later we may refactor the code to just save Generation, or simply get the ClusterSet generation in `updateStatus`

Reply:
Sure, I will go through this part and refine in next release.